### ISSUE TITLE
Can't build without HDF5 on sparse branch?

### ISSUE THREAD

**Issue by Yurlungur** [parthenon-hpc-lab/parthenon#612]
I tried building sparse without HDF5 and got errors. I think we need the ability to build without IO for testing purposes.

**Comment by pgrete** [comment]
I can confirm
```bash
In file included from /home/pgrete/src/athenapk/external/parthenon/src/mesh/mesh.cpp:51:
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:50:21: error: ‘hbool_t’ was not declared in this scope; did you mean ‘bool’?
   50 |     std::unique_ptr<hbool_t[]> allocated;
      |                     ^~~~~~~
      |                     bool
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:50:30: error: template argument 1 is invalid
   50 |     std::unique_ptr<hbool_t[]> allocated;
      |                              ^
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:50:30: error: template argument 2 is invalid
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:71:5: error: ‘hid_t’ does not name a type; did you mean ‘uid_t’?
   71 |     hid_t type;
      |     ^~~~~
      |     uid_t
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:72:5: error: ‘H5D’ does not name a type
   72 |     H5D dataset;
      |     ^~~
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:73:5: error: ‘H5S’ does not name a type
   73 |     H5S dataspace;
      |     ^~~
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:75:5: error: ‘hsize_t’ does not name a type; did you mean ‘ssize_t’?
   75 |     hsize_t count;
      |     ^~~~~~~
      |     ssize_t
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:76:17: error: ‘hsize_t’ was not declared in this scope; did you mean ‘ssize_t’?
   76 |     std::vector<hsize_t> dims;
      |                 ^~~~~~~
      |                 ssize_t
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:76:24: error: template argument 1 is invalid
   76 |     std::vector<hsize_t> dims;
      |                        ^
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:76:24: error: template argument 2 is invalid
In file included from /home/pgrete/src/athenapk/external/parthenon/src/kokkos_abstraction.hpp:30,
                 from /home/pgrete/src/athenapk/external/parthenon/src/parthenon_arrays.hpp:30,
                 from /home/pgrete/src/athenapk/external/parthenon/src/defs.hpp:28,
                 from /home/pgrete/src/athenapk/external/parthenon/src/bvals/bvals_interfaces.hpp:31,
                 from /home/pgrete/src/athenapk/external/parthenon/src/bvals/cc/bvals_cc_in_one.hpp:25,
                 from /home/pgrete/src/athenapk/external/parthenon/src/mesh/mesh.cpp:37:
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp: In member function ‘bool parthenon::RestartReader::SparseInfo::IsAllocated(int, int) const’:
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:56:42: error: invalid operands of types ‘const int’ and ‘std::nullptr_t’ to binary ‘operator!=’
   56 |       PARTHENON_REQUIRE_THROWS(allocated != nullptr,
      |                                ~~~~~~~~~~^~~~~~~~~~
/home/pgrete/src/athenapk/external/parthenon/src/utils/error_checking.hpp:41:9: note: in definition of macro ‘PARTHENON_REQUIRE_THROWS’
   41 |   if (!(condition)) {                                                                    \
      |         ^~~~~~~~~
In file included from /home/pgrete/src/athenapk/external/parthenon/src/mesh/mesh.cpp:51:
/home/pgrete/src/athenapk/external/parthenon/src/outputs/restart.hpp:63:23: error: invalid types ‘const int[int]’ for array subscript
   63 |       return allocated[block * num_sparse + sparse_field_idx];
      |                       ^
make[2]: *** [parthenon/src/CMakeFiles/parthenon.dir/build.make:426: parthenon/src/CMakeFiles/parthenon.dir/mesh/mesh.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1518: parthenon/src/CMakeFiles/parthenon.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

**Comment by Yurlungur** [comment]
I would suggest that when we fix this, we add a regression test to catch this in the future.

**Comment by pgrete** [comment]
agreed

unrelated: I'm still tracing the errors we see with the sparse branch. Interestingly they only show for GPU builds and not for CPU

**Comment by pgrete** [comment]
Fixed with commit d7f03b9 in #586
The fix now defines the unknown hdf5 datatypes, which does the job but is pretty ugly.
I opened #644 so that we don't forget to clean up at some point.
